librec/src/main/java/librec/baseline/ItemCluster.java

```diff
@@ -1,215 +1,215 @@
+//
+package librec.baseline;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+import librec.data.Configuration;
+import librec.data.DenseMatrix;
+import librec.data.DenseVector;
+import librec.data.SparseMatrix;
+import librec.data.SparseVector;
+import librec.data.VectorEntry;
+import librec.intf.GraphicRecommender;
+import librec.util.Logs;
+import librec.util.Randoms;
+
+/**
+ * It is a graphical model that clusters items into K groups for recommendation, as opposite to the {@code UserCluster} recommender.
+ * 
+ * @author Guo Guibing
+ *
+ */
+@Configuration("factors, max.iters")
+public class ItemCluster extends GraphicRecommender {
+
+	private DenseMatrix Pkr; // theta
+	private DenseVector Pi; // pi
+
+	private DenseMatrix Gamma;
+	private DenseMatrix Nir;
+	private DenseVector Ni;
+
+	public ItemCluster(SparseMatrix trainMatrix, SparseMatrix testMatrix, int fold) {
+		super(trainMatrix, testMatrix, fold);
+	}
+
+	@Override
+	protected void initModel() throws Exception {
+
+		Pkr = new DenseMatrix(numFactors, numLevels); // k x r
+		for (int k = 0; k < numFactors; k++) {
+			// random probs 
+			double[] probs = Randoms.randProbs(numLevels);
+			for (int r = 0; r < numLevels; r++)
+				Pkr.set(k, r, probs[r]);
+		}
+
+		Pi = new DenseVector(Randoms.randProbs(numFactors));
+
+		Gamma = new DenseMatrix(numItems, numFactors); // i x k
+
+		// pre-computing Nir, Ni
+		Nir = new DenseMatrix(numItems, numLevels); // Nir
+		Ni = new DenseVector(numItems); // Ni
+
+		for (int i = 0; i < numItems; i++) {
+			SparseVector ri = trainMatrix.column(i);
+
+			for (VectorEntry ve : ri) {
+				double rui = ve.get();
+				int r = ratingScale.indexOf(rui);
+
+				Nir.add(i, r, 1);
+			}
+
+			Ni.set(i, ri.size());
+		}
+	}
+
+	@Override
+	protected void buildModel() throws Exception {
+		for (int iter = 1; iter <= numIters; iter++) {
+
+			// e-step: compute Gamma_ik = P(Zi_k | ri, theta)
+			for (int i = 0; i < numItems; i++) {
+				BigDecimal sum_i = BigDecimal.ZERO;
+				SparseVector ri = trainMatrix.column(i);
+
+				BigDecimal[] sum_ik = new BigDecimal[numFactors];
+				for (int k = 0; k < numFactors; k++) {
+					BigDecimal pik = new BigDecimal(Pi.get(k));
+
+					for (VectorEntry ve : ri) {
+						double rui = ve.get();
+						int r = ratingScale.indexOf(rui);
+						BigDecimal pkr = new BigDecimal(Pkr.get(k, r));
+
+						pik = pik.multiply(pkr);
+					}
+
+					sum_ik[k] = pik;
+					sum_i = sum_i.add(pik);
+				}
+
+				for (int k = 0; k < numFactors; k++) {
+					double zik = sum_ik[k].divide(sum_i, 6, RoundingMode.HALF_UP).doubleValue();
+					Gamma.set(i, k, zik);
+				}
+			}
+
+			// m-step: update Pkr, Pi
+			double sum_ik[] = new double[numFactors];
+			double sum = 0;
+			for (int k = 0; k < numFactors; k++) {
+
+				for (int r = 0; r < numLevels; r++) {
+
+					double numerator = 0, denorminator = 0;
+					for (int i = 0; i < numItems; i++) {
+						double rik = Gamma.get(i, k);

```

Comment:
In order to avoid confusion, the responsibility variable in the E-Step of the model ItemCluster should be named as **rik**, rather than **ruk** as in the model UserCluster.